presto-main/src/test/java/com/facebook/presto/sql/gen/TestCursorProcessorCompiler.java

```diff
@@ -0,0 +1,221 @@
+    private static final SqlFunctionProperties sqlFunctionProperties = SESSION.getSqlFunctionProperties();
+
+    private static final CallExpression ADD_X_Y = call(
+            ADD.name(),
+            FUNCTION_MANAGER.resolveOperator(ADD, fromTypes(BIGINT, BIGINT)),
+            BIGINT,
+            field(0, BIGINT),
+            field(1, BIGINT));
+
+    private static final CallExpression ADD_X_Y_GREATER_THAN_2 = call(
+            GREATER_THAN.name(),
+            FUNCTION_MANAGER.resolveOperator(GREATER_THAN, fromTypes(BIGINT, BIGINT)),
+            BOOLEAN,
+            ADD_X_Y,
+            constant(2L, BIGINT));
+
+    private static final CallExpression ADD_X_Y_LESS_THAN_10 = call(
+            LESS_THAN.name(),
+            FUNCTION_MANAGER.resolveOperator(LESS_THAN, fromTypes(BIGINT, BIGINT)),
+            BOOLEAN,
+            ADD_X_Y,
+            constant(10L, BIGINT));
+
+    private static final CallExpression ADD_X_Y_Z = call(
+            ADD.name(),
+            FUNCTION_MANAGER.resolveOperator(ADD, fromTypes(BIGINT, BIGINT)),
+            BIGINT,
+            call(
+                    ADD.name(),
+                    FUNCTION_MANAGER.resolveOperator(ADD, fromTypes(BIGINT, BIGINT)),
+                    BIGINT,
+                    field(0, BIGINT),
+                    field(1, BIGINT)),
+            field(2, BIGINT));
+
+    @Test
+    public void testRewriteRowExpressionWithCSE()
+    {
+        RowExpression filter = new SpecialFormExpression(AND, BIGINT, ADD_X_Y_GREATER_THAN_2);
+        ImmutableList<RowExpression> rowExpressions = ImmutableList.of(ADD_X_Y_Z, filter);
+        Map<LambdaDefinitionExpression, LambdaBytecodeGenerator.CompiledLambda> compiledLambdaMap = generateMethodsForLambda(cursorProcessorClassDefinition, callSiteBinder, cachedInstanceBinder, rowExpressions, meta, sqlFunctionProperties, "");
+
+        Map<Integer, Map<RowExpression, VariableReferenceExpression>> commonSubExpressionsByLevel = collectCSEByLevel(rowExpressions);
+
+        Map<VariableReferenceExpression, CommonSubExpressionRewriter.CommonSubExpressionFields> cseFields = declareCommonSubExpressionFields(cursorProcessorClassDefinition, commonSubExpressionsByLevel);
+        Map<RowExpression, VariableReferenceExpression> commonSubExpressions = commonSubExpressionsByLevel.values().stream()
+                .flatMap(m -> m.entrySet().stream())
+                .collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
+        // X+Y as CSE
+        assertEquals(1, cseFields.size());
+        VariableReferenceExpression cseVariable = cseFields.keySet().iterator().next();
+
+        RowExpression rewrittenFilter = cseCursorCompiler.rewriteRowExpressionsWithCSE(ImmutableList.of(filter), commonSubExpressions).get(0);
+
+        List<RowExpression> rewrittenProjections = cseCursorCompiler.rewriteRowExpressionsWithCSE(ImmutableList.of(ADD_X_Y_Z), commonSubExpressions);
```

Comment:
`rewriteRowExpressionsWithCSE(rowExpressions, commonSubExpressions)`? You might want to rename `rowExpressions` to `projections` to be clear.